worlds/sc2/rules.py

```diff
@@ -29,6 +35,121 @@ def is_item_placement(self, state):
         # has_group with count = 0 is always true for item placement and always false for SC2 item filtering
         return state.has_group("Missions", self.player, 0)
 
+    # Needs to react on how many missions actually got generated
+    def total_mission_count(self):
+        return 1 if (self.world is None or not hasattr(self.world, 'custom_mission_order')) \
+            else self.world.custom_mission_order.get_mission_count()
+
+    # Unit classes in world, these are set properly into the world after the item culling is done
+    # Therefore need to get from the world
+    def world_has_barracks_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_barracks_unit
+
+    def world_has_factory_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_factory_unit
+
+    def world_has_starport_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_starport_unit
+
+    def world_has_zerg_melee_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_zerg_melee_unit
+
+    def world_has_zerg_ranged_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_zerg_ranged_unit
+
+    def world_has_zerg_air_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_zerg_air_unit
+
+    def world_has_protoss_ground_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_protoss_ground_unit
+
+    def world_has_protoss_air_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_protoss_air_unit
+
+    def weapon_armor_upgrade_count(self, upgrade_item: str, state: CollectionState) -> int:
+        assert upgrade_item in upgrade_bundle_inverted_lookup.keys()
+        count: int = 0
+        if self.generic_upgrade_missions > 0:
+            if not self.is_item_placement(state):
+                # Item pool filtering, W/A upgrades aren't items
+                return WEAPON_ARMOR_UPGRADE_MAX_LEVEL
+            else:
+                count += floor(
+                    (100 / self.generic_upgrade_missions)
+                    * (state.count_group("Missions", self.player) / self.total_mission_count())
+                )
```

Comment:
Is this part of the check desirable? This only runs when the generator is placing items, so the mission layout and item pool are already decided. The only impact I could see is that it would try to place units / upgrades to make more missions beatable _without_ upgrades; with the current rules I think the only recourse it really has is to use ATX instead of starship weapon attack in some cases. Otherwise, this would just cause extra generation failures.

I think it may be more desirable to leave generic upgrade levels as an escape hatch from this logic, so people can not worry about these checks and can set up whatever evil mission order they want?

Reply:
Removing this would not fix #210 as the item placement won't take into account W/A upgrades (and the game would've been unbeatable on Brutal). The mission layout can be even get evil without the player making it actually evil

There are also some failures that relate to Kerrigan levels granted by beaten missions and have the same nature
I think the only way we can intercept that is checking the order in pre-fill and placing some stuff into the start inventory

Reply:
Solved. `pre_fill` will handle these evil cases. Also if you have standard tactics, `generic_upgrade_missions` at 100, you'll have your L3 in the start inventory. No-logic doesn't do that for now (unless we find some issues with keys + no logic)